### PR TITLE
119 multi server maintenance marker

### PR DESCRIPTION
**Pull request by UlrichKu**
| Q             | A
| ------------- | ---
| Branch        | 6.2.x
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes (none added: operates on file level)
| Fixed issues  | chameleon-system/chameleon-system#119
| License       | MIT

### CODE REVIEW

File: src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeServiceInterface.php [hunk 1]
```diff
@@ -15,6 +15,8 @@
 
 interface MaintenanceModeServiceInterface
 {
+    public function isActivated(): bool;
```
Comment (umulmrum) [inline src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeServiceInterface.php]: --> isActive

File: src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeService.php [hunk 2]
```diff
@@ -34,6 +34,17 @@ public function __construct(Connection $connection, CacheInterface $cache)
         $this->cache = $cache;
     }
 
+    public function isActivated(): bool
+    {
+        if (true === \TdbCmsConfig::GetInstance()->fieldShutdownWebsites) {
+            clearstatcache(true, PATH_MAINTENANCE_MODE_MARKER);
```
Comment (umulmrum) [inline src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeService.php]: We should add a comment that describes why we clear the stat cache - I think this is not self-explanatory. And as the maintenance mode functionality is quite scattered across the code base (unavoidably), we should also write some RST documentation describing our thoughts (including multi-node support).

File: src/CoreBundle/EventListener/InitializeRequestListener.php [hunk 3]
```diff
@@ -40,6 +46,31 @@ public function onKernelRequest(GetResponseEvent $event)
...
         $this->requestInitializer->initialize($event->getRequest());
     }
+
+    private function recheckMaintenanceMode(GetResponseEvent $event): void
+    {
+        try {
+            if (true === $this->maintenanceModeService->isActivated()) {
+                $this->redirectToCurrentPage($event);
+            }
+        } catch (MaintenanceModeErrorException $exception) {
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/EventListener/InitializeRequestListener.php]: ok (no exception here anymore anyway)

File: src/CoreBundle/FrontController/chameleon.php [hunk 4]
```diff
@@ -220,6 +221,11 @@ private function showMaintenanceModePage()
         }
     }
 
+    private function renewMaintenanceMarkerFileCache(): void
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/FrontController/chameleon.php]: ok

File: src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeServiceInterface.php (hunk 1 above)
Comment (UlrichKu) [reply src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeServiceInterface.php]: ok

File: src/CoreBundle/EventListener/InitializeRequestListener.php [hunk 5]
```diff
@@ -40,6 +46,31 @@ public function onKernelRequest(GetResponseEvent $event)
...
+    }
+
+    private function redirectToCurrentPage(GetResponseEvent $event): void
+    {
+        $request = $event->getRequest();
+        if (true === $request->isMethodSafe()) {
+            $event->setResponse(new RedirectResponse($_SERVER['REQUEST_URI']));
+        } else {
+            // Redirect is not meaningful for a POST request:
+            $event->setResponse(new Response('Maintenance mode is now active', Response::HTTP_SERVICE_UNAVAILABLE));
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/EventListener/InitializeRequestListener.php]: ok

File: src/CoreBundle/FrontController/chameleon.php [hunk 6]
```diff
@@ -100,6 +100,7 @@ public function boot()
         TTools::AddStaticPageVariables($aStaticURLMapping);
 
         if ($this->isInMaintenanceMode($requestType)) {
+            $this->renewMaintenanceMarkerFileCache();
```
[the remark this answers is not in the document]
Comment (UlrichKu) [reply src/CoreBundle/FrontController/chameleon.php]: ok (done)

But the ticket doesn't specify/require this?

File: src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeService.php (hunk 2 above)
Comment (UlrichKu) [reply src/CoreBundle/Maintenance/MaintenanceMode/MaintenanceModeService.php]: ok ok

File: src/CoreBundle/FrontController/chameleon.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (umulmrum) [reply src/CoreBundle/FrontController/chameleon.php]: Every failed request could mean that a potential customer leaves the site, so if we can avoid it quite easily, we should use this opportunity :-)

**Comment by umulmrum** [comment]
Looks good to me - I reworked the RST a bit, could you have a look if you're OK with the changes?

**Comment by UlrichKu** [comment]
RST changes ok.

**Comment by umulmrum** [comment]
Thanks @UlrichKu

**Pull request merged**